sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala

```diff
@@ -282,8 +315,87 @@ object CatalystTypeConverters {
       }
     }
 
-    override def toScalaImpl(row: InternalRow, column: Int): Row =
-      toScala(row.getStruct(column, structType.size))
+    private[this] def toScalaConverterWithHint: InternalRow => Any = {
+      val className = structToScalaHint.get.erasure.typeSymbol.asClass.fullName
+      val classTag = if (Utils.classIsLoadable(className)) {
+        scala.reflect.ClassTag(Utils.classForName(className))
+      } else {
+        throw new IllegalArgumentException(s"$className is not loadable")
+      }
+
+      StructConverter.createToProductConverter(structType, classTag)
+    }
+  }
+
+  private object StructConverter {
+
+    def createToProductConverter[T <: Product](
+        structType: StructType, classTag: ClassTag[T]): InternalRow => T = {
+
+      import universe._
+
+      val constructorMirror = {
+        val mirror = runtimeMirror(Utils.getContextOrSparkClassLoader)
+        val classSymbol = mirror.classSymbol(classTag.runtimeClass)
+        val classMirror = mirror.reflectClass(classSymbol)
+        val constructorSymbol = {
+          // Adapted from ScalaReflection to find primary constructor.
+          // https://issues.apache.org/jira/browse/SPARK-4791
+          val symbol = classSymbol.toType.declaration(nme.CONSTRUCTOR)
+          if (symbol.isMethod) {
+            symbol.asMethod
+          } else {
+            val candidateSymbol =
+              symbol.asTerm.alternatives.find { s => s.isMethod && s.asMethod.isPrimaryConstructor }
+            if (candidateSymbol.isDefined) {
+              candidateSymbol.get.asMethod
+            } else {
+              throw new IllegalArgumentException(s"No primary constructor for ${symbol.name}")
+            }
+          }
+        }
+        classMirror.reflectConstructor(constructorSymbol)
+      }
+
+      val params = constructorMirror.symbol.paramss.head.toSeq
+      val fields = structType.fields
+      val dataTypes = fields.map { _.dataType }
+      val converters: Seq[Any => Any] = params.zip(dataTypes).map { case (p, dt) =>
+        createToScalaConverter(p.asTerm.typeSignature, dt)
+      }
+
+      (row: InternalRow) => if (row == null) {
+        null.asInstanceOf[T]
+      } else {
+        try {
+          val convertedArgs =
+            converters.zip(row.toSeq(dataTypes)).map { case (converter, arg) => converter(arg) }
+          constructorMirror.apply(convertedArgs: _*).asInstanceOf[T]
+        } catch {
+          case e: IllegalArgumentException => // argument type mismatch
+            throw new ClassCastException(
+              s"Incompatible class ${classTag.runtimeClass.getName}: ${e.getMessage}")
```

Comment:
In the case of trying to store a null value into a primitive type, the error message is kind of confusing: `Caused by: java.lang.ClassCastException: Incompatible class MyClass: null`.  It might be a lot of work to say which part is causing the error, but I wonder if we shouldn't at least mention some of the things that could be wrong here?